feat: skip reconcile for webhook-patched executor fields

Kevinz857 opened the pull request:
## Summary

  When only webhook-patched executor fields change (PriorityClassName, NodeSelector, Tolerations, Affinity, SchedulerName), skip triggering a reconcile since these fields are automatically applied by the mutating webhook when new pods are created.

  This optimization reduces unnecessary application restarts when users update executor scheduling configurations.

  ## Changes

  - Add `isWebhookPatchedFieldsOnlyChange()` to detect webhook-only field changes
  - Skip reconcile and emit event when only webhook fields changed
  - Add comprehensive unit tests for change detection logic

  ## Supported Fields

  The following executor fields can now be updated without triggering a reconcile:
  - `Executor.PriorityClassName`
  - `Executor.NodeSelector`
  - `Executor.Tolerations`
  - `Executor.Affinity`
  - `Executor.SchedulerName`

  ## How It Works

  These fields are patched by the mutating webhook when executor pods are created. When they change:
  1. Already running pods are not affected
  2. New executor pods will automatically use the updated values
  3. No application restart is needed

  ## Test Plan

  - [x] Unit tests for `isWebhookPatchedFieldsOnlyChange()` function
  - [x] Unit tests for `hasExecutorWebhookFieldChanges()` function
  - [x] All existing tests pass

  Partial fix for #2766

ChenYi015 commented:
@Kevinz857 Thanks for your contribution! Since this feature will cause a breaking change, I would like to introduce a feature gate machanism first and add an alpha feature like `PartialRestart` for this feature request and it will default to `false`. Users can choose whether to enable this feature or not. Once this feature turns into GA state, we will enable it by default. WDYT?

Kevinz857 commented:
@ChenYi015  Thanks for your feedback! The feature gate approach sounds reasonable to me.

  I agree that introducing a feature gate mechanism for this change is a good idea, as it:
  1. Allows users to opt-in and test the feature in their environments
  2. Provides a safe rollout path from alpha → beta → GA
  3. Avoids unexpected behavior changes for existing users

  I'll work on implementing a feature gate (e.g., `PartialRestart`) that:
  - Defaults to `false` in alpha stage
  - Guards the service/ingress hot-update logic
  - Can be enabled via controller configuration

  Could you point me to any existing feature gate implementation in the codebase I should follow, or would you prefer I propose a new mechanism?

  Thanks!

Kevinz857 commented:
@ChenYi015  Thanks for the feedback! I've updated this PR to use the feature gate mechanism from #2794.

  Changes made:

  1. Added PartialRestart feature gate in pkg/features/features.go:
    - Default: false
    - PreRelease: Alpha
   
  2. Updated event_filter.go to check the feature gate before skipping reconcile:
  ```go
  if features.Enabled(features.PartialRestart) && f.isWebhookPatchedFieldsOnlyChange(oldApp, newApp) {
      // skip reconcile logic
  }
```

  3. Rebased on latest master (includes #2794)

  Users can enable this feature by setting --feature-gates=PartialRestart=true on the controller.

  Please take another look when you have a chance!
